catalog/ui/catalog-ui-search/src/main/webapp/component/session-timeout-modal/session-timeout.controller.js

```diff
@@ -0,0 +1,94 @@
+* even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+* Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+* is distributed along with this program and can be found at
+* <http://www.gnu.org/licenses/lgpl.html>.
+*
+**/
+/* jshint browser: true */
+/* global define */
+define(['jquery',
+        'underscore',
+        'marionette',
+        'js/wreqr',
+        './lightbox/lightbox.view',
+        './session-timeout.view'
+    ], function($, _, Marionette, wreqr, BlockingLightbox, SessionTimeoutModal) {
+        'use strict';
+        var SessionTimeoutController;
+
+        var _this = null;
+        var idleNoticeDuration = 60000;
+        var idleTimeoutThreshold = 900000;  // Length of inactivity that will trigger user timeout
+                                            // See STIG V-69243
+
+        SessionTimeoutController = Marionette.Controller.extend({
+            initialize: function() {
+                _this = this;
+                _this.initTimer();
+
+                wreqr.vent.on("sessionRenewed", _this.initTimer);
+            },
+
+            initTimer: function() {
+                var timer = null;
+                var start = Date.now();
+                var sessionRenewDate = 0;
+                var idleTimeoutDate = idleTimeoutThreshold - idleNoticeDuration + Date.now();
+
+                function setSessionRenewDate() {
+                    $.get("/services/session/expiry", function(res) {
+                        var msUntilTimeout = parseInt(res);
+                        if (msUntilTimeout === 0)
+                            clearTimeout(timer);
+
+                        var msUntilAutoRenew = Math.max(msUntilTimeout * 0.7, msUntilTimeout - 60000);
+                        sessionRenewDate = Date.now() + msUntilAutoRenew;
+                    });
+                }
+
+                function onInterval() {
+                    var now = Date.now();
+
+                    if (now >= sessionRenewDate) {
+                        $.get("/services/session/renew", function() {
+                            console.log("session renewed");
+                            setSessionRenewDate();
+                        });
+                    }
+
+                    if (now >= idleTimeoutDate) {
+                        $(document).off('keydown mousedown scroll');
+                        clearTimeout(timer);
+
+                        var blockingLightbox = BlockingLightbox.generateNewLightbox();
+                        blockingLightbox.model.updateTitle("Session Expiring");
+                        blockingLightbox.model.open();
+                        blockingLightbox.lightboxContent.show(new SessionTimeoutModal({
+                            time: idleNoticeDuration
+                        }));
+                    }
+                    else
+                        timer = setTimeout(onInterval, 1000 - ((now - start) % 1000));
```

Comment:
There's a problem here - the timer stops whenever the timeout modal is displayed. Initially, this timer only kept track of when the modal should be shown, and this was ok. Now it also has the responsibility of keeping track of when the session will expire on the backend, so the controller can keep it alive.

Now, the controller refreshes a session when there are min {sessionLength * 0.3, 60} seconds left. For a SAML assertion lifespan of < 3.5 minutes or so, there are < 60 seconds left when the controller decides to refresh, and this window could coincide with the 60 seconds where the modal is displayed - i.e. when the timer is stopped. So the session might expire before the user clicks "Continue working", and the session does not get extended.

We need to keep this timer running always in the controller so the session can always be refreshed on the backend. Same in the admin UI